// @remixproject/engine-vscode/lib/filemanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileManagerPlugin = void 0;
const tslib_1 = require("tslib");
const plugin_api_1 = require("@remixproject/plugin-api");
const path_1 = require("path");
const vscode_1 = require("vscode");
const command_1 = require("./command");
function absolutePath(path) {
    const root = vscode_1.workspace.workspaceFolders[0].uri.fsPath;
    // vscode API will never get permission to WriteFile outside of its workspace directory
    if (!root) {
        return path;
    }
    if (path.startsWith(root)) {
        path = path_1.relative(root, path);
    }
    const result = path_1.join(root, path);
    if (!result.startsWith(root)) {
        throw new Error(`Resolved path is should be inside the open workspace : "${root}". Got "${result}`);
    }
    return result;
}
function relativePath(path) {
    const root = vscode_1.workspace.workspaceFolders[0].uri.fsPath;
    // vscode API will never get permission to WriteFile outside of its workspace directory
    if (!root) {
        return path;
    }
    return path_1.relative(root, path);
}
class FileManagerPlugin extends command_1.CommandPlugin {
    constructor() {
        super(plugin_api_1.filSystemProfile);
        // ------------------------------------------
        // Legacy API. To be removed.
        // ------------------------------------------
        this.getFile = this.readFile;
        this.setFile = this.writeFile;
        this.switchFile = this.open;
    }
    /** Open the content of the file in the context (eg: Editor) */
    open(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const absPath = absolutePath(path);
            const uri = vscode_1.Uri.file(absPath);
            return vscode_1.commands.executeCommand('vscode.open', uri);
        });
    }
    /** Set the content of a specific file */
    writeFile(path, data) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const absPath = absolutePath(path);
            const uri = vscode_1.Uri.file(absPath);
            const encoder = new TextEncoder();
            const uint8Array = encoder.encode(data);
            return vscode_1.workspace.fs.writeFile(uri, Uint8Array.from(uint8Array));
        });
    }
    /** Return the content of a specific file */
    readFile(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const absPath = absolutePath(path);
            const uri = vscode_1.Uri.file(absPath);
            return vscode_1.workspace.fs.readFile(uri).then(content => Buffer.from(content).toString("utf-8"));
        });
    }
    /** Remove a file */
    remove(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const absPath = absolutePath(path);
            const uri = vscode_1.Uri.file(absPath);
            return vscode_1.workspace.fs.delete(uri);
        });
    }
    /** Change the path of a file */
    rename(oldPath, newPath) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const source = vscode_1.Uri.file(absolutePath(oldPath));
            const target = vscode_1.Uri.file(absolutePath(newPath));
            return vscode_1.workspace.fs.rename(source, target);
        });
    }
    /** Upsert a file with the content of the source file */
    copyFile(src, dest) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const source = vscode_1.Uri.file(absolutePath(src));
            const target = vscode_1.Uri.file(absolutePath(dest));
            return vscode_1.workspace.fs.copy(source, target);
        });
    }
    /** Create a directory */
    mkdir(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const uri = vscode_1.Uri.file(absolutePath(path));
            return vscode_1.workspace.fs.createDirectory(uri);
        });
    }
    /** Get the list of files in the directory */
    readdir(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const absPath = absolutePath(path);
            const uri = vscode_1.Uri.file(absPath);
            return vscode_1.workspace.fs.readDirectory(uri).then(data => data.map(([path]) => relativePath(path)));
        });
    }
    getCurrentFile() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fileName = vscode_1.window.activeTextEditor ? vscode_1.window.activeTextEditor.document.fileName : undefined;
            return relativePath(fileName);
        });
    }
    /** @deprecated Use readdir instead */
    getFolder(path) {
        throw new Error('Get folder is not supported anymore');
    }
}
exports.FileManagerPlugin = FileManagerPlugin;
