// @remixproject/engine-vscode/lib/webview.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWebview = exports.WebviewPlugin = void 0;
const tslib_1 = require("tslib");
const engine_1 = require("@remixproject/engine");
const vscode_1 = require("vscode");
const path_1 = require("path");
const fs_1 = require("fs");
const https_1 = require("https");
const url_1 = require("url");
class WebviewPlugin extends engine_1.PluginConnector {
    constructor(profile, options) {
        super(profile);
        this.listeners = [];
        this.setOptions(options);
    }
    setOptions(options) {
        super.setOptions(options);
    }
    send(message) {
        var _a;
        (_a = this.panel) === null || _a === void 0 ? void 0 : _a.webview.postMessage(message);
    }
    connect(url) {
        if (this.options.context) {
            this.panel = createWebview(this.profile, url, this.options);
            this.listeners = [
                this.panel.webview.onDidReceiveMessage(msg => this.getMessage(msg)),
                this.panel.onDidDispose(_ => this.call('manager', 'deactivatePlugin', this.name)),
                this.panel,
            ];
        }
        else {
            throw new Error(`WebviewPlugin "${this.name}" `);
        }
    }
    disconnect() {
        this.listeners.forEach(disposable => disposable.dispose());
    }
}
exports.WebviewPlugin = WebviewPlugin;
function isHttpSource(protocol) {
    return protocol === 'https:' || protocol === 'http:';
}
/** Create a webview */
function createWebview(profile, url, options) {
    var _a;
    const { protocol, path } = url_1.parse(url);
    const isRemote = isHttpSource(protocol);
    if (isRemote) {
        return remoteHtml(url, profile, options);
    }
    else {
        const relativeTo = options.relativeTo || 'extension';
        let fullPath;
        if (path_1.isAbsolute(path)) {
            fullPath = path;
        }
        else if (relativeTo === 'extension') {
            const { extensionPath } = options.context;
            fullPath = path_1.join(extensionPath, path);
        }
        else if (relativeTo === 'workspace') {
            const root = (_a = vscode_1.workspace.workspaceFolders[0]) === null || _a === void 0 ? void 0 : _a.uri.fsPath;
            if (!root) {
                throw new Error('No open workspace. Cannot find url of relative path: ' + path);
            }
            fullPath = path_1.join(root, path);
        }
        return localHtml(fullPath, profile, options);
    }
}
exports.createWebview = createWebview;
///////////////
// LOCAL URL //
///////////////
/** Create panel webview based on local HTML source */
function localHtml(url, profile, options) {
    var _a;
    const { ext } = path_1.parse(url);
    const baseUrl = ext === '.html' ? path_1.parse(url).dir : url;
    const panel = vscode_1.window.createWebviewPanel(profile.name, profile.displayName || profile.name, options.column || ((_a = vscode_1.window.activeTextEditor) === null || _a === void 0 ? void 0 : _a.viewColumn) || vscode_1.ViewColumn.One, {
        enableScripts: true,
        localResourceRoots: [vscode_1.Uri.file(baseUrl)]
    });
    setLocalHtml(panel.webview, baseUrl);
    // Devmode
    if (options.devMode) {
        const index = path_1.join(baseUrl, 'index.html');
        fs_1.watch(index).on('change', _ => setLocalHtml(panel.webview, baseUrl));
    }
    return panel;
}
/** Get code from local source */
function setLocalHtml(webview, baseUrl) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const index = `${baseUrl}/index.html`;
        // Get all links from "src" & "href"
        const matchLinks = /(href|src)="([^"]*)"/g;
        // Vscode requires URI format from the extension root to work
        const toUri = (original, prefix, link) => {
            // For: <base href="#" /> && remote url : <link href="https://cdn..."/>
            const isRemote = isHttpSource(url_1.parse(link).protocol);
            if (link === '#' || isRemote) {
                return original;
            }
            // For scripts & links
            const path = path_1.join(baseUrl, link);
            const uri = vscode_1.Uri.file(path);
            return `${prefix}="${webview['asWebviewUri'](uri)}"`;
        };
        const html = yield fs_1.promises.readFile(index, 'utf-8');
        webview.html = html.replace(matchLinks, toUri);
    });
}
////////////////
// REMOTE URL //
////////////////
/** Create panel webview based on remote HTML source */
function remoteHtml(url, profile, options) {
    var _a;
    const { ext } = path_1.parse(url);
    const baseUrl = ext === '.html' ? path_1.parse(url).dir : url;
    const panel = vscode_1.window.createWebviewPanel(profile.name, profile.displayName || profile.name, options.column || ((_a = vscode_1.window.activeTextEditor) === null || _a === void 0 ? void 0 : _a.viewColumn) || vscode_1.ViewColumn.One, { enableScripts: true });
    setRemoteHtml(panel.webview, baseUrl);
    return panel;
}
/** Fetch remote ressource with http */
function fetch(url) {
    return new Promise((resolve, reject) => {
        https_1.get(url, res => {
            let text = '';
            res.on('data', data => text += data);
            res.on('end', _ => resolve(text));
            res.on('error', err => reject(err));
        });
    });
}
/** Get code from remote source */
function setRemoteHtml(webview, baseUrl) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const matchLinks = /(href|src)="([^"]*)"/g;
        const index = `${baseUrl}/index.html`;
        // Vscode requires URI format from the extension root to work
        const toRemoteUrl = (original, prefix, link) => {
            // For: <base href="#" /> && remote url : <link href="https://cdn..."/>
            const isRemote = isHttpSource(url_1.parse(link).protocol);
            if (link === '#' || isRemote) {
                return original;
            }
            // For scripts & links
            const path = path_1.join(baseUrl, link);
            return `${prefix}="${path}"`;
        };
        const html = yield fetch(index);
        webview.html = html.replace(matchLinks, toRemoteUrl);
    });
}
