// @remixproject/engine-web/lib/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemePlugin = exports.createTheme = void 0;
const tslib_1 = require("tslib");
const engine_1 = require("@remixproject/engine");
const plugin_api_1 = require("@remixproject/plugin-api");
/**
 * Utils function to create a theme with default value
 * Default values are taken from material design with colors
 * - primary: indigo
 * - secondary: pink
 * - warn: orange
 * - error: red
 */
function createTheme(params = {}) {
    return Object.assign(Object.assign({ brightness: 'light', fontFamily: 'Arial,"Helvetica Neue",Helvetica,sans-serif', space: 8 }, params), { breakpoints: Object.assign({ xs: 0, sm: 600, md: 1024, lg: 1440, xl: 1920 }, params.breakpoints), colors: Object.assign({ surface: 'white', background: '#fafafa', foreground: 'black', primary: '#3f51b5', primaryContrast: 'white', secondary: '#e91e63', secondaryContrast: 'rgba(white, 0.7)', success: '#4caf50', successContrast: 'rgba(black, 0.87)', warn: '#ff9800', warnContrast: 'white', error: '#f44336', errorContrast: 'white', disabled: 'rgba(0,0,0,.26)' }, params.colors) });
}
exports.createTheme = createTheme;
class ThemePlugin extends engine_1.Plugin {
    constructor() {
        super(plugin_api_1.themeProfile);
        this.getTheme = createTheme;
        this.theme = this.getTheme();
    }
    /** Internal API to set the current theme */
    setTheme(theme) {
        this.theme = this.getTheme(theme);
        this.emit('themeChanged', this.theme);
    }
    /** External API to get the current theme */
    currentTheme() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.theme;
        });
    }
}
exports.ThemePlugin = ThemePlugin;
