// @remixproject/engine-web/lib/web-worker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebWorkerPlugin = void 0;
const tslib_1 = require("tslib");
const engine_1 = require("@remixproject/engine");
class WebWorkerPlugin extends engine_1.PluginConnector {
    constructor(profile, options) {
        super(profile);
        this.profile = profile;
        this.setOptions(options);
    }
    setOptions(options) {
        super.setOptions(Object.assign({ type: 'module', name: this.name }, options));
    }
    connect(url) {
        if ('Worker' in window) {
            this.worker = new Worker(url, this.options);
            this.worker.onmessage = e => this.getEvent(e);
            return this.handshake();
        }
    }
    disconnect() {
        this.worker.terminate();
    }
    /** Get message from the iframe */
    getEvent(event) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const message = event.data;
            this.getMessage(message);
        });
    }
    /**
     * Post a message to the webview of this plugin
     * @param message The message to post
     */
    send(message) {
        if (!this.worker) {
            throw new Error('No worker attached to the plugin');
        }
        this.worker.postMessage(message);
    }
}
exports.WebWorkerPlugin = WebWorkerPlugin;
