// @remixproject/engine-web/lib/ws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebsocketPlugin = void 0;
const tslib_1 = require("tslib");
const engine_1 = require("@remixproject/engine");
class WebsocketPlugin extends engine_1.PluginConnector {
    constructor(profile, options = {}) {
        super(profile);
        // Listener is needed to remove the listener
        this.listeners = {
            message: ['message', (e) => this.getEvent(e), false],
            close: ['close', (e) => this.onclose(e), false],
        };
        this.options = {
            reconnectDelay: 1000
        };
        this.options = Object.assign(Object.assign({}, this.options), options);
    }
    getEvent(event) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const message = JSON.parse(event.data);
            this.getMessage(message);
        });
    }
    /** Try to reconnect to net websocket if closes */
    onclose(e) {
        this.loaded = false;
        // Abnormal closing Try to reconnect
        if (e.code !== 1000) {
            setTimeout(() => this.open(), this.options.reconnectDelay);
        }
    }
    /** Open a connection with the server (also used for reconnection) */
    open() {
        this.socket = new WebSocket(this.url);
        this.socket.addEventListener('open', () => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.socket.addEventListener(...this.listeners.message);
            this.handshake();
        }));
    }
    send(message) {
        if (this.socket.readyState !== this.socket.OPEN) {
            throw new Error('Websocket connection is not open yet');
        }
        this.socket.send(JSON.stringify(message));
    }
    connect(url) {
        this.url = url;
        this.open();
        this.socket.addEventListener(...this.listeners.close);
    }
    disconnect() {
        this.socket.removeEventListener(...this.listeners.close);
        this.socket.removeEventListener(...this.listeners.message);
        this.socket.close();
    }
}
exports.WebsocketPlugin = WebsocketPlugin;
